onnxscript/rewriter/ort_fusions/fuse_mha_bias.py

```diff
@@ -0,0 +1,203 @@
+        query_matmul,
+        key_matmul,
+        value_matmul,
+        q_bias,
+        k_bias,
+        v_bias,
+        mask,
+        past_key,
+        past_value,
+        num_heads,
+    ):
+        if not self._q_no_bias:
+            query_BSD = op.Add(query_matmul, q_bias)
+        else:
+            query_BSD = query_matmul
+        if not self._k_no_bias:
+            key_BSD = op.Add(key_matmul, k_bias)
+        else:
+            key_BSD = key_matmul
+        if not self._v_no_bias:
+            value_BSD = op.Add(value_matmul, v_bias)
+        else:
+            value_BSD = value_matmul
+
+        return op.MultiHeadAttention(
+            query_BSD,
+            key_BSD,
+            value_BSD,
+            None,  # bias
+            None,  # key padding mask
+            mask,  # attention mask/bias
+            past_key,
+            past_value,
+            num_heads=num_heads,
+            _domain="com.microsoft",
+        )
+
+    def check(
+        self,
+        op,
+        query_matmul,
+        key_matmul,
+        value_matmul,
+        **_,
+    ) -> pattern.MatchResult:  # type: ignore[name-defined]
+        check_result = pattern.MatchResult()
+
+        self.bindings: dict[str, Dim] = {}
+
+        def no_match(val: ir.Value, dims: Sequence[str]) -> bool:
+            return not _fusion_utils._check_shape(self.bindings, val, dims)
+
+        if no_match(query_matmul, ["B", "S", "D"]):
+            return check_result.fail(
+                f"Shape mismatch: {query_matmul} does not match expected dimensions ['B', 'S', 'D']",
+                query_matmul,
+            )
+        if no_match(key_matmul, ["B", "Skv", "Dk"]):
+            return check_result.fail(
+                f"Shape mismatch: {key_matmul} does not match expected dimensions ['B', 'Skv', 'Dk']",
+                key_matmul,
+            )
+        if no_match(value_matmul, ["B", "Skv", "Dv"]):
+            return check_result.fail(
+                f"Shape mismatch: {value_matmul} does not match expected dimensions ['B', 'Skv', 'Dv']",
+                value_matmul,
+            )
+
+        self.Dh_q = self.bindings.get("D")
+        self.Dh_k = self.bindings.get("Dk")
+        self.Dh_v = self.bindings.get("Dv")
+
+        if (
+            not isinstance(self.Dh_q, int)
+            or not isinstance(self.Dh_k, int)
+            or not isinstance(self.Dh_v, int)
+        ):
+            return check_result.fail(
+                "Could not determine the hidden sizes of query, key, and value.",
+            )
+
+        return check_result
+
+    def rewrite(
+        self,
+        op,
+        query_matmul,
+        key_matmul,
+        value_matmul,
+        q_bias,
+        k_bias,
+        v_bias,
+        mask,
+        past_key,
+        past_value,
+        num_heads,
+        **_,
+    ):
+        if self._q_no_bias:
+            q_bias = op.Constant(
+                value=ir.tensor(numpy.zeros((self.Dh_q,), dtype=numpy.float32))
```

Comment:
Do we care only about 32-bit float query/key/value?

Reply:
Modified to have dtype of the input to the mha, the contrib op only supports float and float16, so that will be the constraint on the bias as well